mips/atomic.c

```diff
@@ -0,0 +1,56 @@
+#include <atomic.h>
+
+#define __MIPS_PLATFORM_SYNC_NOPS ""
+static __inline void mips_sync(void) {
+  __asm __volatile(".set noreorder\n"
+                   "\tsync\n" __MIPS_PLATFORM_SYNC_NOPS ".set reorder\n"
+                   :
+                   :
+                   : "memory");
+}
+
+static __inline uint32_t atomic_cmpset_32(__volatile uint32_t *p,
```

Comment:
GCC gives a number of different synchronization primitives described [here](https://gcc.gnu.org/onlinedocs/gcc-5.4.0/gcc/_005f_005fsync-Builtins.html#g_t_005f_005fsync-Builtins). I've checked our new toolchain and indeed `__sync_bool_compare_and_swap` is defined in `libgcc.a`. The disassembly gives following code:

```
00000000 <__sync_bool_compare_and_swap_4>:
   0:   0000000f        sync
   4:   c0830000        ll      v1,0(a0)
   8:   14650006        bne     v1,a1,24 <__sync_bool_compare_and_swap_4+0x24>
   c:   24020000        li      v0,0
  10:   00c00821        move    at,a2
  14:   e0810000        sc      at,0(a0)
  18:   1020fffa        beqz    at,4 <__sync_bool_compare_and_swap_4+0x4>
  1c:   24020001        li      v0,1
  20:   0000000f        sync
  24:   03e00008        jr      ra
  28:   00000000        nop
```